### DISCUSSION TITLE
can't mock import.meta in vi.hoisted

### DISCUSSION THREAD

**Discussion by HomyeeKing** [vitest-dev/vitest#5307] [Q&A]
### Describe the bug

```ts
vi.hoisted(() => {
  import.meta.a = 1; // not work
  global.ddd = 2 // works
  return {
    rdt: vi.fn(),
  };
});
```

### Reproduction

https://github.com/HomyeeKing/vitest-repo/tree/importMeta

### System Info

```shell
System:
    OS: macOS 13.6
    CPU: (8) arm64 Apple M1
    Memory: 76.45 MB / 16.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 16.19.0 - ~/.nvm/versions/node/v16.19.0/bin/node
    Yarn: 1.22.17 - ~/.yarn/bin/yarn
    npm: 9.9.2 - ~/.nvm/versions/node/v16.19.0/bin/npm
  Browsers:
    Chrome: 122.0.6261.69
    Safari: 16.6
```

### Used Package Manager

pnpm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by hi-ogawa** [comment]
I'm not too familiar with spec, but I think `import.meta` is an independent object for each file and it's not something expected to be shared globally within the runtime like `globalThis`.

> The import.meta meta-property exposes context-specific metadata to a JavaScript module. 

from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import.meta

Maybe can you explain your use case how you're actually trying to use `import.meta`? That might help finding an alternative.

**Comment by HomyeeKing** [comment]
@hi-ogawa   
```ts
// const.ts
export const isServer = import.meta.render === 'server'
export const isServer = import.meta.render === 'client'

// foo.ts
import {isServer} from './const'
export function foo(){
  if(isServer){
// server code
  }else{
  //  client code
	}
}
```

```ts
// foo.test.ts
import {foo} from './foo'
test('foo',()=>{
	foo()
})
```

First, I want to test code in `server` case  or  in `client` , I have alternative that mock `./const`
```ts
vi.mock('./const', async (importOriginal) => {
  const mod: object = await importOriginal();
  return {
    ...mod,
    isServer: true // or isClient: true,
  };
});
```

Second, I hope to change `import.meta.render`（or other variable） in **one test** file dynamicly,  or change specific setupFile for some test file ( no relative API so far)

**Reply by hi-ogawa** [reply]
I think I understand what you're trying to do, but there's a still missing context.
How is `import.meta.render` usually defined, for example, when you're running app? I think normally `import.meta.render` would be `undefined` unless you setup something in the plugin or somewhere.

For example, if you have a control over `import.meta.render`, then you could change it to `import.meta.env.render`, which I think would be shared between all files, so `vi.hoisted` idea might work in that case.

**Reply by HomyeeKing** [reply]
it's defined by the framework I am using, I can do nothing with it. So `vi.mock` is the alternative for this case?

I'd also like to discuss the second question(maybe in another thread, I just write here for now):
forget about the `import.meta`, like if get value from `globalThis.render`, I want to change `globalThis.render`  as I want.

like set `globalThis.render = 'server'` firstly to test server code, then change it to `client` by some API  like `vi.setupFile('client-setup.ts')` or `vi.hoisted.set(key, value)`  to do the test again with the new setupFile

**Reply by hi-ogawa** [reply]
> `vi.mock` is the alternative for this case?

Yeah, this looks okay to me.

> like set `globalThis.render = 'server'` firstly to test server code, then change it to `client` by some API 

Since each test files are "isolated" (unless you have explicitly set `isolated: false`), I think you can modify `globalThis` as you like inside `vi.hoisted(() => globalThis.render = "server")` and that shouldn't affect other tests.

**Reply by HomyeeKing** [reply]
OK, just confirm one thing that can only do this by split two files and set `vi.hoisted` respectly
```ts
✅
// foo.server.test.ts
vi.hoisted(() => globalThis.render = "server")
test('server',()={
	// test server code
})


// foo.client.test.ts
vi.hoisted(() => globalThis.render = "client")
test('client',()={
	// test client code
})
```
NOT  in one file (I felt this way is a bit unreasonable when I was writing, since hoist is hoist, thoungh it can reuse some common code)
```ts
❌
// foo.test.ts
const mocks = vi.hoisted(() => globalThis.render = "server")
test('server',()={
	// test server code
})

test('client',()=>{
     mocks.update(()=>{
		//  update globalThis here again
	})
	// test client code
})
```

**Reply by hi-ogawa** [reply]
> NOT in one file

Yeah, supporting that use case sounds unrealistic and I feel it's more like an architectural issue of your `const.ts` system if you really want to make that work in a single test file.

**Reply by HomyeeKing** [reply]
got it, thank you!
